torch_geometric/explain/algorithm/gnn_explainer.py

```diff
@@ -0,0 +1,536 @@
+        num_nodes = x.size(0)
+        num_edges = edge_index.size(1)
+        x, edge_index, index, subset, hard_edge_mask, kwargs =\
+            self.subgraph(model, index, x, edge_index, **kwargs)
+        if target_index is not None and model_config.mode ==\
+                ModelMode.classification:
+            target = torch.index_select(target, 1, subset)
+        else:
+            target = target[subset]
+
+        self._train_node_edge_mask(model, x, edge_index, explainer_config,
+                                   model_config, target, target_index, index,
+                                   **kwargs)
+
+        if explainer_config.node_mask_type == MaskType.common_attributes:
+            node_mask = self.node_mask.detach().sigmoid().squeeze(0)
+            node_mask = self._reshape_common_attributes(node_mask, num_nodes)
+        else:
+            if explainer_config.node_mask_type == MaskType.object:
+                new_mask = x.new_zeros(num_nodes, 1)
+            if explainer_config.node_mask_type == MaskType.attributes:
+                new_mask = x.new_zeros(num_nodes, x.size(-1))
+            new_mask[subset] = self.node_mask.detach().sigmoid()
+            node_mask = new_mask.squeeze(-1)
+
+        if explainer_config.edge_mask_type == MaskType.object:
+            new_edge_mask = torch.zeros(num_edges)
+            new_edge_mask[hard_edge_mask] = self.edge_mask.detach().sigmoid()
+            edge_mask = new_edge_mask
+        else:
+            edge_mask = None
+
+        return node_mask, edge_mask
+
+    def _train_node_edge_mask(
+        self,
+        model: torch.nn.Module,
+        x: Tensor,
+        edge_index: Tensor,
+        explainer_config: ExplainerConfig,
+        model_config: ModelConfig,
+        target: Tensor,
+        target_index: Optional[Union[int, Tensor]] = None,
+        index: Optional[Union[int, Tensor]] = None,
+        **kwargs,
+    ):
+        self._initialize_masks(x, edge_index,
+                               node_mask_type=explainer_config.node_mask_type,
+                               edge_mask_type=explainer_config.edge_mask_type)
+
+        if explainer_config.edge_mask_type == MaskType.object:
+            set_masks(model, self.edge_mask, edge_index, apply_sigmoid=True)
+            parameters = [self.node_mask, self.edge_mask]
+        else:
+            parameters = [self.node_mask]
+        optimizer = torch.optim.Adam(parameters, lr=self.lr)
+
+        for _ in range(self.epochs):
+            optimizer.zero_grad()
+            h = x * self.node_mask.sigmoid()
+            out = model(x=h, edge_index=edge_index, **kwargs)
+            loss_value = self.loss(
+                out, target, return_type=model_config.return_type,
+                target_idx=target_index, node_index=index,
+                edge_mask_type=explainer_config.edge_mask_type,
+                model_mode=model_config.mode)
+            loss_value.backward(retain_graph=True)
```

Comment:
I had an issue computing the gradient otherwise, the error indicated I should add that. Could not figure out why, but I checked in the examples the explainer is learning sensible masks.

Reply:
Could you paste the error here, lets try to resolve it. 
Setting `retain_graph` to `True` is pretty inefficient.

Reply:
If you set `retrain_graph=True` and run the tests, you will get (for example) 

```
_____________________ test_gnn_explainer_with_meta_explainer_regression_node[phenomenon-raw-GCN_multioutput_regression-common_attributes-None] _____________________

edge_mask_type = None, node_mask_type = 'common_attributes', model = GCN_multioutput_regression(
  (conv1): GCNConv(3, 16)
  (conv2): GCNConv(16, 7)
)
return_type = 'raw', explanation_type = 'phenomenon'

    @pytest.mark.parametrize('edge_mask_type', edge_mask_types)
    @pytest.mark.parametrize('node_mask_type', node_mask_types)
    @pytest.mark.parametrize(
        'model', [GCN_single_output_regression, GCN_multioutput_regression])
    @pytest.mark.parametrize('return_type', return_types_regression)
    @pytest.mark.parametrize("explanation_type", ["model", "phenomenon"])
    def test_gnn_explainer_with_meta_explainer_regression_node(
        edge_mask_type,
        node_mask_type,
        model,
        return_type,
        explanation_type,
    ):
        explainer_config = ExplainerConfig(
            explanation_type=explanation_type,
            node_mask_type=node_mask_type,
            edge_mask_type=edge_mask_type,
        )
        model_config = ModelConfig(
            task_level='node',
            return_type=return_type,
            mode="regression",
        )
    
        model = model()
    
        out = model(x, edge_index)
    
        if isinstance(model, GCN_multioutput_regression):
            target_index = 0
        else:
            target_index = None
    
        explainer = Explainer(model=model, algorithm=GNNExplainer(),
                              explainer_config=explainer_config,
                              model_config=model_config)
    
        # try to explain prediction for node 0
>       explanation = explainer(x=x, edge_index=edge_index, target=out,
                                target_index=target_index, node_index=2)

test/explain/algoritms/test_gnn_explainer.py:378: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
torch_geometric/explain/explainer.py:115: in __call__
    explanation = self.algorithm(
env/lib/python3.9/site-packages/torch/nn/modules/module.py:1130: in _call_impl
    return forward_call(*input, **kwargs)
torch_geometric/explain/algorithm/gnn_explainer.py:98: in forward
    node_mask, edge_mask = self._explain_node(model, x, edge_index,
torch_geometric/explain/algorithm/gnn_explainer.py:172: in _explain_node
    self._train_node_edge_mask(model, x, edge_index, explainer_config,
torch_geometric/explain/algorithm/gnn_explainer.py:228: in _train_node_edge_mask
    loss_value.backward(retain_graph=False)
env/lib/python3.9/site-packages/torch/_tensor.py:396: in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

tensors = (tensor([[0.6374]], grad_fn=<AddBackward0>),), grad_tensors = None, retain_graph = False, create_graph = False, grad_variables = None, inputs = ()

    def backward(
        tensors: _TensorOrTensors,
        grad_tensors: Optional[_TensorOrTensors] = None,
        retain_graph: Optional[bool] = None,
        create_graph: bool = False,
        grad_variables: Optional[_TensorOrTensors] = None,
        inputs: Optional[_TensorOrTensors] = None,
    ) -> None:
        r"""Computes the sum of gradients of given tensors with respect to graph
        leaves.
    
        The graph is differentiated using the chain rule. If any of ``tensors``
        are non-scalar (i.e. their data has more than one element) and require
        gradient, then the Jacobian-vector product would be computed, in this
        case the function additionally requires specifying ``grad_tensors``.
        It should be a sequence of matching length, that contains the "vector"
        in the Jacobian-vector product, usually the gradient of the differentiated
        function w.r.t. corresponding tensors (``None`` is an acceptable value for
        all tensors that don't need gradient tensors).
    
        This function accumulates gradients in the leaves - you might need to zero
        ``.grad`` attributes or set them to ``None`` before calling it.
        See :ref:`Default gradient layouts<default-grad-layouts>`
        for details on the memory layout of accumulated gradients.
    
        .. note::
            Using this method with ``create_graph=True`` will create a reference cycle
            between the parameter and its gradient which can cause a memory leak.
            We recommend using ``autograd.grad`` when creating the graph to avoid this.
            If you have to use this function, make sure to reset the ``.grad`` fields of your
            parameters to ``None`` after use to break the cycle and avoid the leak.
    
        .. note::
    
            If you run any forward ops, create ``grad_tensors``, and/or call ``backward``
            in a user-specified CUDA stream context, see
            :ref:`Stream semantics of backward passes<bwd-cuda-stream-semantics>`.
    
        .. note::
    
            When ``inputs`` are provided and a given input is not a leaf,
            the current implementation will call its grad_fn (even though it is not strictly needed to get this gradients).
            It is an implementation detail on which the user should not rely.
            See https://github.com/pytorch/pytorch/pull/60521#issuecomment-867061780 for more details.
    
        Args:
            tensors (Sequence[Tensor] or Tensor): Tensors of which the derivative will be
                computed.
            grad_tensors (Sequence[Tensor or None] or Tensor, optional): The "vector" in
                the Jacobian-vector product, usually gradients w.r.t. each element of
                corresponding tensors. None values can be specified for scalar Tensors or
                ones that don't require grad. If a None value would be acceptable for all
                grad_tensors, then this argument is optional.
            retain_graph (bool, optional): If ``False``, the graph used to compute the grad
                will be freed. Note that in nearly all cases setting this option to ``True``
                is not needed and often can be worked around in a much more efficient
                way. Defaults to the value of ``create_graph``.
            create_graph (bool, optional): If ``True``, graph of the derivative will
                be constructed, allowing to compute higher order derivative products.
                Defaults to ``False``.
            inputs (Sequence[Tensor] or Tensor, optional): Inputs w.r.t. which the gradient
                be will accumulated into ``.grad``. All other Tensors will be ignored. If
                not provided, the gradient is accumulated into all the leaf Tensors that
                were used to compute the attr::tensors.
        """
        if grad_variables is not None:
            warnings.warn("'grad_variables' is deprecated. Use 'grad_tensors' instead.")
            if grad_tensors is None:
                grad_tensors = grad_variables
            else:
                raise RuntimeError("'grad_tensors' and 'grad_variables' (deprecated) "
                                   "arguments both passed to backward(). Please only "
                                   "use 'grad_tensors'.")
        if inputs is not None and len(inputs) == 0:
            raise RuntimeError("'inputs' argument to backward() cannot be empty.")
    
        tensors = (tensors,) if isinstance(tensors, torch.Tensor) else tuple(tensors)
        inputs = (inputs,) if isinstance(inputs, torch.Tensor) else \
            tuple(inputs) if inputs is not None else tuple()
    
        grad_tensors_ = _tensor_or_tensors_to_tuple(grad_tensors, len(tensors))
        grad_tensors_ = _make_grads(tensors, grad_tensors_, is_grads_batched=False)
        if retain_graph is None:
            retain_graph = create_graph
    
        # The reason we repeat same the comment below is that
        # some Python versions print out the first line of a multi-line function
        # calls in the traceback and some print out the last line
>       Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
            tensors, grad_tensors_, retain_graph, create_graph, inputs,
            allow_unreachable=True, accumulate_grad=True)  # Calls into the C++ engine to run the backward pass
E       RuntimeError: Trying to backward through the graph a second time (or directly access saved tensors after they have already been freed). Saved intermediate values of the graph are freed when you call .backward() or autograd.grad(). Specify retain_graph=True if you need to backward through the graph a second time or if you need to access saved tensors after calling backward.

env/lib/python3.9/site-packages/torch/autograd/__init__.py:173: RuntimeError

Reply:
Do all tests give you this error, also do the examples give you this error?  Will try running this on my end too.
This error usually means something isn't working as expected.

Reply:
Found the issue: in the test I was calling the model to compute the ground_truth without a `torch.no_grad()`. For the classification case this went unnoticed as I had the `argmax` operator was breaking the gradients anyway. 
This has no impact on the implementation, as the Explainer class correctly sets a context manager `with torch.no_grad()`

Reply:
Good catch.